task_processing/interfaces/event.py

```diff
@@ -5,24 +5,49 @@
 from pyrsistent import freeze
 from pyrsistent import m
 from pyrsistent import PMap
+from pyrsistent import pmap
 from pyrsistent import PRecord
 
 
+EVENT_KINDS = ['task', 'control']
+
+
 class Event(PRecord):
+    kind = field(type=str,
+                 invariant=lambda x: (x in EVENT_KINDS,
+                                      'kind not in {}'.format(EVENT_KINDS)))
     # we store timestamps as seconds since epoch.
     # use time.time() to generate
     timestamp = field(type=float)
     # reference to platform-specific event object
     raw = field()
+    # free-form dictionary for stack-specific data
+    extensions = field(type=PMap, initial=m(), factory=pmap)
     # is this the last event for a task?
     terminal = field(type=bool, mandatory=True)
-    success = field(type=(bool, type(None)), initial=None)
-    # platform-specific event name
-    platform_type = field(type=str)
+
+    # task-specific fields
     # task_id this event pertains to
     task_id = field(type=str)
-    task_config = field()
-    extensions = field(type=PMap, initial=m())
+    # task config dict that sourced the task this event refers to
+    task_config = field(type=PMap, factory=pmap)
+    # the task finished with exit code 0
+    success = field(type=(bool, type(None)), initial=None)
+    # platform-specific event type
+    platform_type = field(type=str)
+
+    # control events
+    message = field(type=str)
+
+
+def task_event(**kwargs):
```

Comment:
Why is this better than just having separate PRecords for TaskEvent and ControlEvent? I know inheritance hierarchies are sad but it's not that bad when you have type hierarchies imo.